docs/user/build-customization.rst

```diff
@@ -0,0 +1,164 @@
+
+
+These jobs can be declared by using a :doc:`/config-file/index` with the :ref:`config-file/v2:build.jobs` key on it.
+Let's say the project requires commands to be executed *before* installing any dependency into the Python environment and *after* the build has finished.
+In that case, a config file similar to this one can be used:
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-20.04"
+     tools:
+       python: "3.10"
+     jobs:
+       pre_install:
+         - bash ./scripts/pre_install.sh
+       post_build:
+         - curl -X POST \
+           -F "project=${READTHEDOCS_PROJECT}" \
+           -F "version=${READTHEDOCS_VERSION}" https://example.com/webhooks/readthedocs/
+
+
+There are some caveats to knowing when using user-defined jobs:
+
+* ``PWD`` environment variable is the root of the cloned repository
+* Environment variables are expanded in the commands (see :doc:`environment-variables`)
+* Each command is executed in a new shell process, so modifications done to the shell environment are not persistent between commands
+* Any command returning non-zero exit code makes the build to fail immediately
+* ``build.os`` and ``build.tools`` are required when using ``build.jobs``
+
+
+Examples
+--------
+
+We've included some common examples where using :ref:`config-file/v2:build.jobs` will be useful.
+These examples may require some adaptation for each projects' use case,
+we recommend you use them as a starting point.
+
+
+Unshallow clone
+~~~~~~~~~~~~~~~
+
+Read the Docs does not perform a full clone on ``checkout`` job to reduce network data and speed up the build process.
+Because of this, extensions that depend on the full Git history will fail.
+To avoid this, it's possible to unshallow the clone done by Read the Docs:
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-20.04"
+     tools:
+       python: "3.10"
+     jobs:
+       post_checkout:
+         - git fetch --unshallow
+
+
+Generate documentation from annotated sources with Doxygen
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It's possible to run Doxygen as part of the build process to generate documentation from annotated sources:
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-20.04"
+     tools:
+       python: "3.10"
+     jobs:
+       pre_build:
+         - doxygen
+
+
+Use MkDocs extensions with extra required steps
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There are some MkDocs extensions that require specific commands to be run to generate extra pages before performing the build.
+For example, `pydoc-markdown <http://niklasrosenstein.github.io/pydoc-markdown/>`_
+
+.. code:: yaml
+
+   version: 2
+   build:
+     os: "ubuntu-20.04"
+     tools:
+       python: "3.10"
+     jobs:
+       pre_build:
+         - pydoc-markdown --build --site-dir "$PWD/_build/html"
+
+
+Avoid having a dirty ``git`` index
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
To match other instances of "Git", but also, keep literal strings out of headings/toc

```suggestion
Avoid having a dirty Git index
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```